### PR TITLE
Add track()

### PR DESCRIPTION
**Pull request by maksimf**

### CODE REVIEW

**Comment by patricioe** [comment]
remember to issue against development branch

**Comment by maksimf** [comment]
fixed

**Comment by patricioe** [comment]
@maksimf how are the events flush being trigger? I don't see that in your PR.

File: lib/splitclient-rb/cache/adapters/redis_adapter.rb [hunk 1]
```diff
@@ -118,6 +118,23 @@ def random_set_elements(key, count)
           @redis.srandmember(key, count)
         end
 
+        # Queue
+        def add_to_queue(key, val, max_size)
+          @redis.del(key) if @redis.llen(key) >= max_size
```
Comment (sarrubia) [inline lib/splitclient-rb/cache/adapters/redis_adapter.rb]: @maksimf my concern here is that:  
 - if this delete key when max_size is reached is executed in redis-shared mode, the events into redis LIST will be removed and the consumer (split synchronizer) will lost those deleted events.

I think that eventsQueueSize configuration param is only to in-memory mode (not redis)

**Comment by maksimf** [comment]
@patricioe this PR is `[WIP]`

File: lib/splitclient-rb/cache/adapters/redis_adapter.rb (hunk 1 above)
Comment (maksimf) [reply lib/splitclient-rb/cache/adapters/redis_adapter.rb]: ok, thank you, I'll fix that

**Comment by maksimf** [comment]
@patricioe please take a look

File: NEWS [hunk 2]
```diff
@@ -1,3 +1,7 @@
+4.4.0
+
+Add track() API
```
Comment (patricioe) [inline NEWS]: replace with "Add the ability to send event data to split service (.track() api)

File: lib/splitclient-rb/cache/repositories/events/memory_repository.rb [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+          def initialize(adapter, config)
+            @adapter = adapter
+            @config = config
+          end
+
+          def add(key, traffic_type, event_type, time, value)
+            @adapter.add_to_queue(m: metadata, e: event(key, traffic_type, event_type, time, value))
+          rescue ThreadError # queue is full
+            @config.logger.warn("Dropping events. Current size is #{@config.events_queue_size}. " \
```
Comment (patricioe) [inline lib/splitclient-rb/cache/repositories/events/memory_repository.rb]: can this be logged only if debug is enabled?

File: lib/splitclient-rb/cache/senders/events_sender.rb [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+        end
+
+        def call
+          if ENV['SPLITCLIENT_ENV'] == 'test'
+            post_events
+          else
+            events_thread
+
+            if defined?(PhusionPassenger)
+              PhusionPassenger.on_event(:starting_worker_process) do |forked|
```
Comment (patricioe) [inline lib/splitclient-rb/cache/senders/events_sender.rb]: what's the story with puma server here?

Comment (maksimf) [reply lib/splitclient-rb/cache/senders/events_sender.rb]: Same thing as with every other thread we have: we're recreating threads for Passenger workers

File: lib/splitclient-rb/cache/adapters/redis_adapter.rb [hunk 5]
```diff
@@ -118,6 +118,21 @@ def random_set_elements(key, count)
           @redis.srandmember(key, count)
         end
 
+        # Queue
+        def add_to_queue(key, val)
+          @redis.rpush(key, val)
+        end
+
+        def get_from_queue(key, count)
```
Comment (sarrubia) [inline lib/splitclient-rb/cache/adapters/redis_adapter.rb]: LGTM

**Comment by maksimf** [comment]
@patricioe fixed

**Pull request merged**